redisuniversal/redis.go

```diff
@@ -0,0 +1,229 @@
+package redisuniversal
+
+import (
+"encoding/json"
+"strings"
+"fmt"
+"os"
+"os/signal"
+"time"
+
+"github.com/go-redis/redis"
+"gopkg.in/launchdarkly/go-server-sdk.v4"
+"gopkg.in/launchdarkly/go-server-sdk.v4/ldlog"
+"gopkg.in/launchdarkly/go-server-sdk.v4/utils"
+)
+
+type redisStoreCore struct {
+	loggers ldlog.Loggers
+	config  Options
+	client  redis.UniversalClient
+	clusterMode bool
+}
+
+const hashTag = "{ld}."
+
+func newRedisStoreCore(options Options, loggers ldlog.Loggers) *redisStoreCore {
+	loggers.SetPrefix("RedisFeatureStore:")
+	client := redis.NewUniversalClient(options.RedisOpts)
+
+	// ping the server so we know we are good
+	err := client.Ping().Err()
+	if err != nil {
+		loggers.Error("could not ping redis server: %v", err)
+		return nil
+	}
+
+	c := make(chan os.Signal)
+	signal.Notify(c, os.Interrupt, os.Kill)
+	go func() {
+		select {
+		case <-c:
+			_ = client.Close()
+		}
+	}()
+	if options.MaxRetryCount == 0 {
+		options.MaxRetryCount = defaultRetryCount
+	}
+	if options.CacheTTL == 0 {
+		options.CacheTTL = DefaultCacheTTL
+	}
+	var clusterMode bool
+	if len(options.RedisOpts.Addrs) > 1 {
+		clusterMode = true
+	}
+
+	return &redisStoreCore{
+		loggers: loggers,
+		client:  client,
+		config:  options,
+		clusterMode: clusterMode,
+	}
+}
+func (c *redisStoreCore) featuresKey(kind ldclient.VersionedDataKind) string {
+	return c.hashTagKey(c.config.Prefix + ":" + kind.GetNamespace())
+}
+
+func (c *redisStoreCore) initedKey() string {
+	return c.config.Prefix + ":" + initedKey
+}
+
+// We use a hashtag in order to keep all keys in the same node (and hash slot) so we can perform
+// and use watch ... exec without issues. Only in ClusterMode
+func (c *redisStoreCore) hashTagKey(key string) string {
+	if c.clusterMode {
+		return hashTag + key
+	}
+	return key
+}
+
+// Removing the hashtag from the key
+func (c *redisStoreCore) cleanHashTagKey(key string) string {
+	if c.clusterMode {
+		return strings.Replace(key, hashTag, "",-1)
```

Comment:
I think `TrimPrefix` would have been more appropriate since the operation you're trying to reverse here was specifically adding `hashTag` to the _beginning_ of the string.